File: Math&Geometry/3307. Find the K-th Character in String Game II/solution.py
from math import ceil, log2

class Solution:
    def kthCharacter(self, k: int, operations: List[int]) -> str:
        # n = 1
        # op = 0
        # while n < k:
        #     n *= 2
        #     op += 1
        op = ceil(log2(k))
        n = pow(2, op)

        shift = 0
        for i in range(op-1, -1, -1):
            if k > n//2: # 代表是透過操作得到的
                if operations[i] == 1:
                    shift = (shift+1)%26

                k = k-n//2
            # else:
            #     do nothing

            n //= 2
        
        return chr(ord("a") + shift)

File: Math&Geometry/3307. Find the K-th Character in String Game II/solution.cpp
class Solution {
public:
    char kthCharacter(long long k, vector<int>& operations) {
        long long size = 1;
        int total_op = 0;
        while (size < k) {
            total_op++;
            size *= 2;
        }

        int shift = 0;
        for (int i=total_op-1; i>=0; i--) {
            if (k > size/2) /* from shift&append operation */ {
                if (operations[i] == 1) /* shift&append */ {
                    shift++;
                    k -= size/2;
                } else /* copy */ {
                    k -= size/2;
                }
            }
            size /= 2;
        }
        return 'a' + (shift%26);
    }
};